Improve Reconciliation

gabrielhartmann opened the pull request:
We now do explicit and implicit reconciliation.  This improves startup
time because we no longer blindly wait 30s hoping reconciliation
completes.  This also fixes a correctness problem.  We were always
transitioning out of the Reconciliation acquisition phase regardless of
the status updates we received after 30s.  Now we don't exit
reconciliation until we receive all the needed status updates.

There are also design improvements: moving the reconciliation
implementation out of the Scheduler class, and using the Observer model
to receive status update notifications.

Example output from a reconciliation session is available here:
https://docs.google.com/document/d/1dCe-7LFnqsg_hZB3As31yGG2jkDMIrqlrp5wmmMp0jE/edit

kensipe commented:
in general LGTM

I like that we are pulling code apart and separating out the reconciliation is great.
There are few nits which are that big of a deal... 
1. I was hoping Scheduler would have less responsibility not more.  I would love for it not to be runnable..  In my mind there would be a separate thread responsible for that.   In the case of Observable, It seems necessary.
2.  I'm a minimalist when it comes to code... so line 29 of Reconciler has "extra" unnecessary cruft <String> is redundant java 1.7+
3. While I might return quickly in a method to avoid the body of the method if certain conditions are not met, I don't like too many or unnecessary returns (it leads to complicated code)... so line 109 of LiveState is something I don't like... I would rather have an else on the rest of the code (since that is what it is :)


That said... the code is clear and easy to read.  We reconcile at registration and on status updates... very easy to understand IMO.  nice work.

kensipe commented:
has this been tested?

gabrielhartmann commented:
@kensipe: Thanks for the comments, I'll make the recommended code changes.  I have tested killing the Scheduler and am currently testing killing off nodes and will speak to @elingg about any other needed tests.  I won't merge until everyone's happy this has been fully tested.

kensipe commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java:
```diff
@@ -0,0 +1,110 @@
+
+  public Reconciler(IPersistentStateStore pStore) {
+    store = pStore;
+    pendingTasks = new HashSet<String>();
```
HashSet<String> could be HashSet.   It is assigned to a Set<String> so <String> is implied... and enforced by the compiler.

gabrielhartmann replied:
"HashSet could be HashSet?"  Did you mean "HashSet could be Set" or "Set could be HashSet"?  Set is abstract right, so I thought I needed a real implementation here.

kensipe replied:
It looks like my comment changed after I submitted which made things confusing... what I meant was that `pendingTasks = new HashSet<String>();` could be `pendingTasks = new HashSet();`

The `<>` must be filtered on GH which makes sense.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java:
```diff
@@ -0,0 +1,110 @@
+
+  public void reconcile(SchedulerDriver driver) {
+    pendingTasks = store.getAllTaskIds();
+    logPendingTasks();
```
Hm, we're not updating our persistent state here so I'm not convinced reconciliation still works. I don't see the PersistentState or LIveState being updated unless I'm missing something here. Can you test the scenario of shutting down the scheduler killing say a journal node. Then when it reconciles does it realize the journal node is gone and update the state accordingly?

gabrielhartmann replied:
The PersistentState and LiveState are still updated as always.  I have not modified that; it's still done in exactly the same way in the HdfsScheduler.  Reconciliation only consists in asking for StatusUpdates of all persisted Tasks and then implicitly asking for the status of all tasks.  The consequences of receiving those StatusUpdates are the same as always.  The LiveState and PersistentState are updated as always.  I have tested the scenario you suggest, and all works as expected.

elingg replied:
Thanks very much for the explanation. It makes sense, assuming this has been tested. Let's test a few scenarios here. Shutdown down the scheduler. Take down a NN by killing the executor process. Start the scheduler see if recovers. Same for DN and JN. If these basic tests pass, we should be good!

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/HdfsScheduler.java:
```diff
@@ -127,6 +126,10 @@ public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
       status.getMessage(),
       liveState.getStagingTasksSize()));
 
```
Not convinced of the observer pattern here. I don't think we should notify observers every time we get a status update. This means observers will get notified everytime (not just during reconciliation). At least we could notify observers only when we are reconciling.

gabrielhartmann replied:
1) Moving Observer notification there wouldn't work.  The Reconciler must receive updates of TerminalState Tasks so that it can remove them from its list of pendingTasks (for reconciliation).  If it is reconciling a Task and we get a notification that it was LOST we still need that status update.  That's a reconciled task.

2) The idea going forward is that a large portion of the logic in the statusUpdate method should be moved out of the HdfsScheduler class anyway.  Components which need to take action based on the statusUpdate stream are Observers.  They should behave as the Reconciler does here.  When reconciliation is complete, the Reconciler essentially does a no-op when receiving a statusUpdate.

3) Finally statusUpdates are actually pretty infrequent and small in number.  The overhead of sending updates to "all observers" is very low.

elingg replied:
Ok, we discussed in person and reached a conclusion.

gabrielhartmann replied:
I've added an if block to only notify the Reconciler of StatusUpdates when it is is reconciling.  In the future this can be modified if LiveState and PersistentState etc. become Observers as well.

elingg replied:
Cool, thanks. I think this check can be (liveState.getCurrentAcquisitionPhase() == AcquisitionPhase.RECONCILING_TASKS) . We might put the other phase correction check before this one also.

gabrielhartmann replied:
@elingg: This pattern of having the Observervable determine what information the Observers should receive is a bad idea.  This is emblematic of one of the pervasive problems in the over all implementation of HdfsScheduler.  It is constantly inspecting and interacting with the internal state of other Classes, in particular LiveState and PersistentState.  I'm trying to get away from this mistake with the Observer pattern.

The motivation for limiting the information provided to the Observers is also committing the sin of premature optimization.  We're talking about tens of events here and are worried that handling them (in memory) is going to slow things down.  It's a non-issue and shouldn't be a motivation for inspecting the internal state of Observers.

In my view wer're degrading cohesion and breaking any sense of layering for no clear win.

elingg replied:
The scheduler should be able to modify state. I wouldn't consider that a sin. It is making calls to update the state. Then this functionality is actually executed by the state stores. To me, the worst sin of the Scheduler is that there is too much logic there. We need to spend some time removing this tech debt and breaking down the functionality into smaller pieces.

Regarding observers, I don't want to unnecessarily make any calls that aren't required. If the observers ignore these unnecessary calls and don't execute additional logic (other than logging), that is okay with me. As you mentioned these are only a few calls so this is not a deal breaker. The main point here is to be very careful about what logic we want to execute and not make unnecessary calls.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/HdfsScheduler.java:
```diff
@@ -199,6 +202,10 @@ public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
   public void resourceOffers(SchedulerDriver driver, List<Offer> offers) {
     log.info(String.format("Received %d offers", offers.size()));
 
```
Let's check liveState.getCurrentAcquisitionPhase() == AcquisitionPhase.RECONCILING_TASKS before reconciler.complete. Swap the values in your && statement. This is a nit, but it's nice to not enter logic unless we must.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/HdfsScheduler.java:
```diff
@@ -619,6 +626,11 @@ private void reloadConfigsOnAllRunningTasks(SchedulerDriver driver) {
   }
 
   private void correctCurrentPhase() {
+    if (!reconciler.complete()) {
```
I don't think we need this logic here. We should start in reconciling phase and only leave once it's done.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java:
```diff
@@ -0,0 +1,110 @@
+      }
+    }
+  }
+
```
Should this method be called receivedStatusUpdate?

gabrielhartmann replied:
It must be called update.  That's the implementation of the Observer contract.

elingg replied:
Ok, sure. Since we are using the observer pattern here.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java:
```diff
@@ -0,0 +1,110 @@
+      if (taskIsPending(taskId)) {
+        log.info(String.format("Reconciling Task '%s'.", taskId));
+        pendingTasks.remove(taskId);
+      } else {
```
Maybe we should log here: "Already reconciled task '%s'"

gabrielhartmann replied:
It wasn't necessarily already reconciled.  This may be something else altogether.  It could potentially be a notification about the Status of a Task which we did not persist information about.  All I know for sure is that it's a Task that doesn't address any of the Tasks pending reconciliation.

elingg replied:
Perhaps we should say "Task %s has already been reconciled or is unknown."

gabrielhartmann commented:
@elingg: "Shutdown down the scheduler. Take down a NN by killing the executor process. Start the scheduler see if recovers. Same for DN and JN. If these basic tests pass, we should be good!"

I tested all these scenarios and found one bug and fixed it.  It looks like it can happen that the PersistentStateStore can sometimes fail and return a null list when asked for all the TaskIDs.  That's bad and will need to be looked into.  I've made reconciliation fall back to implicit reconciliation in the case when explicit reconciliation is impossible.

elingg commented:
LGTM, one nit comment. Once @kensipe is okay, we can merge.

kensipe commented:
@gabrielhartmann LGTM.  lets merge.

gabrielhartmann commented:
@elingg: I've tested again with the latest version represented here and addressed all comments.  I'm waiting on a response from you on the topic of filtering the status even stream from the Observable side instead of the Observer side.  Other than that it's ready to merge.

elingg commented:
hi @gabrielhartmann, I addressed your comment above. This is not a deal breaker, whether we choose to include the if statement to notify the observers there or not with this patch. If this is tested on DCOS and Mesos, there is nothing blocking you from shipping it. :+1:

gabrielhartmann commented:
@elingg and @kensipe:  Apologies for the re-review request.  If you could both have a quick look at the changes I've made to the LiveState I'd appreciate it.  In this new reconciliation implementation I was able to hit a defect more frequently than in the original implementation.  Since we're doing both explicit and implicit reconciliation, we are getting two status updates for every Task.

When I was testing failing over the scheduler and a NN simultaneously, the two updates for the surviving NN were being interpreted as an update for each NN.  So we were inappropriately skipping the FORMAT_NAME_NODES acquisition phase.  The changes to LiveState guarantee that two updates for the same NN only affect that single NN.  This faulting situation was possible in the old implementation, but much less likely.

elingg commented:
Ah, this makes more sense now @gabrielhartmann. Good thing you caught the fact that the failure was occurring more often with these changes.

I took a look at your code and I would like to see you  simplify the logic in a couple of ways 1) Don't make all the methods public - only if required 2) Have some sensible default values for the NameNodeState- don't include as many null checks. 

Let's think how we can achieve the same purpose in less code. I do like you breaking down this into a separate object as needed.

gabrielhartmann commented:
1) What methods would you make private?  All the currently exposed methods are things which were being accessed directly through Map structure that the NameNodeState replaces.  If I make anything private it will require reworking logic elsewhere which I explicitly wish to avoid.  I think there is plenty of room for improvement in LiveState and NameNode executor, but I'm not sure this is the PR to do it in.

2) It's important that we signal in some way that the TaskStatus has not been set or a particular NN.  At least that is the model which has so far been adopted in LiveState.  This was done previously through checking whether the old Map object was empty.  I think null is the sensible default to signal that something hasn't been set.

Regarding the size of the LiveState change, I'm not sure what you mean.  It's almost a line by line replacement of what was here already doing manipulations of a Map object.  It's a net gain of 10 lines.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -151,4 +151,60 @@ public boolean apply(String taskId) {
+      log.info(String.format("Set Node '%s' state to status: '%s', initialized: '%s'", name, status, initialized));
+    }
+
+    public void clear() {
```
This method is internal so it should be private.

gabrielhartmann replied:
It used to be called externally.  I'll make it private.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -151,4 +151,60 @@ public boolean apply(String taskId) {
+    public void clear(TaskID taskId) {
+      if (status == null) {
+        return;
+      } else {
```
To have a lot less lines of code just call if(hasStatus) No need for an if and an else

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -151,4 +151,60 @@ public boolean apply(String taskId) {
+
+    public NameNodeState(String name) {
+      this.name = name;
+      clear();
```
Why would we need to clear within the constructor. This is when we are creating the object so everything should be clear already.

elingg commented:
I left a few notes in the code on how we can simplify the logic (i.e. no need for a lot of null checks when we can use less code to call an existing method) and how certain methods can be private. Please let me know when you are ready for the rereview.

gabrielhartmann commented:
@elingg: Code golf complete.  Ready re-re-re-review.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -151,4 +151,50 @@ public boolean apply(String taskId) {
+      log.info(String.format("Set Node '%s' state to status: '%s', initialized: '%s'", name, status, initialized));
+    }
+
+    public void clear(TaskID taskId) {
```
let's have this return a boolean true or false if the task was cleared successfully

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/state/LiveState.java:
```diff
@@ -54,13 +56,10 @@ public void removeStagingTask(final Protos.TaskID taskID) {
-      nameNode2TaskMap.clear();
-    }
+    log.info(String.format("Removing running task: %s", taskId));
+
```
Let's only call the second clear method if the first does not return true

gabrielhartmann replied:
Why?  Stylistic preference?  Or is the micro-optimization of calls?  We'd save what a hasStatus() null check call or two if we do it your way.

elingg replied:
Clarity in the code so i know which state is being cleared

elingg commented:
LGTM if it's tested on DCOS and Mesos. One small comment above, but not blocking the merge.

gabrielhartmann commented:
Tested on both DCOS and Mesos.
